gems/smithy-client/lib/smithy-client/net_http/handler.rb

```diff
@@ -18,66 +18,70 @@ def initialize(bytes_expected, bytes_received)
+        def transmit(config, req, resp)
           # Monkey patch default content-type set by Net::HTTP
           Thread.current[:net_http_skip_default_content_type] = true
           session(config, req) do |http|
             http.request(build_net_request(req)) do |net_resp|
-              bytes_received = unpack_response(net_resp, resp)
-              complete_response(req, resp, output, bytes_received)
+              bytes_received = signal_response(net_resp, resp)
+              complete_response(req, resp, bytes_received)
             end
           end
-        rescue ArgumentError
+        rescue ArgumentError => e
           # Invalid verb, ArgumentError is a StandardError
           # Not retryable.
-          raise
+          resp.signal_error(e)
         rescue StandardError => e
-          output.error = NetworkingError.new(e)
+          error = NetworkingError.new(e)
+          resp.signal_error(error)
         ensure
           # ensure we turn off monkey patch in case of error
           Thread.current[:net_http_skip_default_content_type] = nil
         end
 
-        # @param [Net::HTTPResponse] net_resp
-        # @param [HTTP::Response] resp
-        def unpack_response(net_resp, resp)
-          resp.status_code = net_resp.code.to_i
-          net_resp.each_header { |k, v| resp.headers[k] = v }
+        def signal_response(net_resp, resp)
+          status_code = net_resp.code.to_i
+          headers = extract_headers(net_resp)
+
           bytes_received = 0
+          resp.signal_headers(status_code, headers)
           net_resp.read_body do |chunk|
             bytes_received += chunk.bytesize
-            resp.body.write(chunk)
+            resp.signal_data(chunk)
           end
           bytes_received
         end
 
-        def complete_response(req, resp, output, bytes_received)
-          resp.body.rewind if resp.body.respond_to?(:rewind)
-          verify_bytes_received(resp, output, bytes_received) if should_verify_bytes?(req, resp)
+        def complete_response(req, resp, bytes_received)
+          if should_verify_bytes?(req, resp)
+            verify_bytes_received(resp, bytes_received)
+          else
+            resp.signal_done
+          end
         end
 
         def should_verify_bytes?(req, resp)
           req.http_method != 'HEAD' && resp.headers['content-length']
         end
 
-        def verify_bytes_received(resp, output, bytes_received)
+        def verify_bytes_received(resp, bytes_received)
           bytes_expected = resp.headers['content-length'].to_i
-          return if bytes_expected == bytes_received
-
-          error = TruncatedBodyError.new(bytes_expected, bytes_received)
-          output.error = NetworkingError.new(error)
+          if bytes_expected == bytes_received
+            resp.signal_done
+          else
+            error = TruncatedBodyError.new(bytes_expected, bytes_received)
+            resp.signal_error(NetworkingError.new(error))
```

Comment:
Why are we wrapping a truncated body error in another networking error?

Reply:
It's a type of networking error, which wraps networking errors from net http.

Reply:
I guess maybe more what I'm getting at is, why isnt our TruncatedBodyError a NetworkingError?

Reply:
It is a networking error. Networking errors wrap other errors, so we just made our own, with a helpful message. I could make it inherit from NetworkingError but then I would have to change the initializer anyway.